src/materialized/tests/tls.rs

```diff
@@ -0,0 +1,743 @@
+    pub fn request_cert<I>(&self, name: &str, ips: I) -> Result<(PathBuf, PathBuf), Box<dyn Error>>
+    where
+        I: IntoIterator<Item = IpAddr>,
+    {
+        let rsa = Rsa::generate(2048)?;
+        let pkey = PKey::from_rsa(rsa)?;
+        let subject_name = {
+            let mut builder = X509NameBuilder::new()?;
+            builder.append_entry_by_nid(Nid::COMMONNAME, name)?;
+            builder.build()
+        };
+        let cert = {
+            let mut builder = X509::builder()?;
+            builder.set_version(2)?;
+            builder.set_pubkey(&pkey)?;
+            builder.set_issuer_name(&self.cert.subject_name())?;
+            builder.set_subject_name(&subject_name)?;
+            builder.set_not_before(&*Asn1Time::days_from_now(0)?)?;
+            builder.set_not_after(&*Asn1Time::days_from_now(365)?)?;
+            for ip in ips {
+                builder.append_extension(
+                    SubjectAlternativeName::new()
+                        .ip(&ip.to_string())
+                        .build(&builder.x509v3_context(None, None))?,
+                )?;
+            }
+            builder.sign(&self.pkey, MessageDigest::sha256())?;
+            builder.build()
+        };
+        let cert_path = self.dir.path().join(Path::new(name).with_extension("crt"));
+        let key_path = self.dir.path().join(Path::new(name).with_extension("key"));
+        fs::write(&cert_path, &cert.to_pem()?)?;
+        fs::write(&key_path, &pkey.private_key_to_pem_pkcs8()?)?;
+        Ok((cert_path, key_path))
+    }
+}
+
+fn make_pg_tls<F>(configure: F) -> MakeTlsConnector
+where
+    F: Fn(&mut SslConnectorBuilder) -> Result<(), ErrorStack>,
+{
+    let mut connector_builder = SslConnector::builder(SslMethod::tls()).unwrap();
+    configure(&mut connector_builder).unwrap();
+    MakeTlsConnector::new(connector_builder.build())
+}
+
+fn make_http_tls<F>(configure: F) -> HttpsConnector<HttpConnector>
+where
+    F: Fn(&mut SslConnectorBuilder) -> Result<(), ErrorStack>,
+{
+    let mut connector_builder = SslConnector::builder(SslMethod::tls()).unwrap();
+    configure(&mut connector_builder).unwrap();
+    let mut http = HttpConnector::new();
+    http.enforce_http(false);
+    HttpsConnector::with_connector(http, connector_builder).unwrap()
+}
+
+enum Assert<E> {
+    Success,
+    Err(E),
+}
+
+enum TestCase<'a> {
+    Pgwire {
+        user: &'static str,
+        ssl_mode: SslMode,
+        configure: Box<dyn Fn(&mut SslConnectorBuilder) -> Result<(), ErrorStack> + 'a>,
+        assert: Assert<Box<dyn Fn(postgres::Error) + 'a>>,
+    },
+    Http {
+        user: &'static str,
+        scheme: Scheme,
+        configure: Box<dyn Fn(&mut SslConnectorBuilder) -> Result<(), ErrorStack> + 'a>,
+        assert: Assert<Box<dyn Fn(Option<StatusCode>, String) + 'a>>,
+    },
+}
+
+fn run_tests<'a>(header: &str, server: &util::Server, tests: &[TestCase<'a>]) {
+    println!("==> {}", header);
```

Comment:
Intentional actually; this is my hacky recreation of Go's subtests. It's friggin impossible to tell where one table-driven test case starts and ends otherwise. :/